group-coordinator/src/test/java/org/apache/kafka/coordinator/group/OffsetMetadataManagerTest.java

```diff
@@ -0,0 +1,1113 @@
+                        .setName("bar")
+                        .setPartitions(Arrays.asList(
+                            new OffsetCommitResponseData.OffsetCommitResponsePartition()
+                                .setPartitionIndex(0)
+                                .setErrorCode(Errors.OFFSET_METADATA_TOO_LARGE.code()),
+                            new OffsetCommitResponseData.OffsetCommitResponsePartition()
+                                .setPartitionIndex(1)
+                                .setErrorCode(Errors.NONE.code())
+                        ))
+                )),
+            result.response()
+        );
+
+        assertEquals(
+            Collections.singletonList(RecordHelpers.newOffsetCommitRecord(
+                "foo",
+                "bar",
+                1,
+                new OffsetAndMetadata(
+                    100L,
+                    OptionalInt.of(10),
+                    "small",
+                    context.time.milliseconds(),
+                    OptionalLong.empty()
+                ),
+                MetadataImage.EMPTY.features().metadataVersion()
+            )),
+            result.records()
+        );
+    }
+
+    @Test
+    public void testReplay() {
+        OffsetMetadataManagerTestContext context = new OffsetMetadataManagerTestContext.Builder().build();
+
+        verifyReplay(context, "foo", "bar", 0, new OffsetAndMetadata(
+            100L,
+            OptionalInt.empty(),
+            "small",
+            context.time.milliseconds(),
+            OptionalLong.empty()
+        ));
+
+        verifyReplay(context, "foo", "bar", 0, new OffsetAndMetadata(
+            200L,
+            OptionalInt.of(10),
+            "small",
+            context.time.milliseconds(),
+            OptionalLong.empty()
+        ));
+
+        verifyReplay(context, "foo", "bar", 1, new OffsetAndMetadata(
+            200L,
+            OptionalInt.of(10),
+            "small",
+            context.time.milliseconds(),
+            OptionalLong.empty()
+        ));
+
+        verifyReplay(context, "foo", "bar", 1, new OffsetAndMetadata(
+            300L,
+            OptionalInt.of(10),
+            "small",
+            context.time.milliseconds(),
+            OptionalLong.of(12345L)
+        ));
+    }
+
+    @Test
+    public void testReplayWithTombstone() {
+        OffsetMetadataManagerTestContext context = new OffsetMetadataManagerTestContext.Builder().build();
+
+        verifyReplay(context, "foo", "bar", 0, new OffsetAndMetadata(
+            100L,
+            OptionalInt.empty(),
+            "small",
+            context.time.milliseconds(),
+            OptionalLong.empty()
+        ));
+
+        context.replay(RecordHelpers.newOffsetCommitTombstoneRecord(
+            "foo",
+            "bar",
+            0
+        ));
+
+        assertNull(context.offsetMetadataManager.offset("foo", new TopicPartition("bar", 0)));
```

Comment:
maybe we can add some comments on this test that verifyReplay puts the group into the offset map and the tombtone removes it.